File: stud/zhilin/Lab3/task_3.1/main.py
import math


def f(x: float) -> float:
    return math.acos(x)


def lagrange_interpolation(x: float, cords: list[tuple[float, float]]) -> float:
    coefficients, n = [cords[i][1] for i in range(len(cords))], len(cords)
    for i in range(n):
        for j in range(n):
            if i != j:
                coefficients[i] /= cords[i][0]-cords[j][0]
    for i in range(n):
        for j in range(n):
            if i != j:
                coefficients[i] *= x-cords[j][0]
    return sum(coefficients)


def newton_interpolation(x: float, cords: list[tuple[float, float]]) -> float:
    coefficients, n = [cords[i][1] for i in range(len(cords))], len(cords)
    for i in range(1, n):
        for j in range(n-1, i-1, -1):
            coefficients[j] = (coefficients[j]-coefficients[j-1])/(cords[j][0]-cords[j-i][0])
            
    for i in range(1, n):
        for j in range(i):
            coefficients[i] *= x-cords[j][0]
    return sum(coefficients)


if __name__ == '__main__':
    x_vect = [-0.4, -0.1, 0.2, 0.5]
    x_marked = 0.1
    cord = [(x, f(x)) for x in x_vect]

    print('\nМногочлен Лагранжа')
    print(f'\tРезультат: {lagrange_interpolation(x_marked, cord)}')
    print(f'\tОшибка: {abs(lagrange_interpolation(x_marked, cord) - f(x_marked))}')
    print()

    print('Многочлен Ньютона')
    print(f'\tРезультат: {newton_interpolation(x_marked, cord)}')
    print(f'\tОшибка: {abs(newton_interpolation(x_marked, cord) - f(x_marked))}')
    print()

File: stud/zhilin/Lab3/task_3.1/main.cpp
#include <bits/stdc++.h>

using namespace std;

double f(double x) {
    return acos(x);
}

double lagrange_interpolation(double x, const vector<pair<double, double>>& coords) {
    vector<double> coefficients(coords.size());
    int n = coords.size();

    for (int i = 0; i < n; ++i)
        coefficients[i] = coords[i].second;

    for (int i = 0; i < n; ++i)
        for (int j = 0; j < n; ++j)
            if (i != j)
                coefficients[i] /= coords[i].first - coords[j].first;

    for (int i = 0; i < n; ++i)
        for (int j = 0; j < n; ++j)
            if (i != j)
                coefficients[i] *= x - coords[j].first;

    return accumulate(coefficients.begin(), coefficients.end(), 0.0);
}

double newton_interpolation(double x, const vector<pair<double, double>>& coords) {
    vector<double> coefficients(coords.size());
    int n = coords.size();

    for (int i = 0; i < n; ++i)
        coefficients[i] = coords[i].second;

    for (int i = 1; i < n; ++i)
        for (int j = n - 1; j > i-1; --j) 
            coefficients[j] = (coefficients[j] - coefficients[j - 1]) / (coords[j].first - coords[j - i].first);

    for (int i = 1; i < n; ++i) {
        for (int j = 0; j < i; ++j) {
            coefficients[i] *= x - coords[j].first;
        }
    }

    return accumulate(coefficients.begin(), coefficients.end(), 0.0);
}

int main() {
    vector<double> x_vect = {-0.4, -0.1, 0.2, 0.5};
    double x_marked = 0.1;
    vector<pair<double, double>> cord;

    for (double x : x_vect)
        cord.emplace_back(x, f(x));

    cout << "\nThe Lagrange polynomial\n";
    cout << "\tResult: " << lagrange_interpolation(x_marked, cord) << endl;
    cout << "\tLoss: " << abs(lagrange_interpolation(x_marked, cord) - f(x_marked)) << endl;
    cout << "\nThe Newton polynomial\n";
    cout << "\tResult: " << newton_interpolation(x_marked, cord) << endl;
    cout << "\tLoss: " << abs(newton_interpolation(x_marked, cord) - f(x_marked)) << endl;

    return 0;
}
